### DISCUSSION TITLE
Allow `foo()` as syntax sugar for `foo.Invoke()` on any type

### DISCUSSION THREAD

**Discussion by Richiban** [dotnet/csharplang#7084] [Language Ideas]
Delegates, funcs etc have this `Invoke` method on them that allows you to execute the delegate, but you can also just directly call the delegate with `()` which amounts to the same thing. This makes it feel like `()` is just syntax sugar for `.Invoke()`, and it's always bothered me that you can't do this with regular classes or interfaces. With my code style often taking a codebase to the point of mostly single-method interfaces, it's often impossible to give that single method a meaningful name and it ends up just being called `Execute()` or `Handle()` or something else equally meaningless.

Allowing classes and interfaces to have a method called `Invoke` that doesn't have to be written at the callsite would help with understanding in these cases, making these single-method interfaces _feel like functions_, enabling a more functional code style. 

E.g.

```cs
interface IGetCustomerQuery
{
    Customer Invoke(CustomerId id);
}

public class MyService(IGetCustomerQuery getCustomer)
{
    public void M()
    {
        var customer = getCustomer(new CustomerId(123));
        
        //...
    }
}
```

------

Edit: I'll add this about precedence: both Kotlin and Scala have `foo(args)` as syntax sugar for `foo.apply(args)`.

**Comment by iam3yal** [comment]
Related #95

**Comment by TahirAhmadov** [comment]
Could it be beneficial to decorate the method (or the type) with an attribute to indicate which method to use? Imagine all the existing code with the `Handle` and `Execute` and `Perform` methods (I have quite a few such types) and renaming all of those to `Invoke` will be difficult. Also, what about `HandleAsync` and the like?

**Reply by iam3yal** [reply]
@TahirAhmadov Wouldn't something like this work? `void Invoke() => Execute()` whatever the signature and method is so you don't really have to rename anything and it just becomes a syntactic sugar for existing code I also don't think that it fits all scenarios sometimes it makes more sense and more readable to use the actual name and some people to this  day prefer to use `.Invoke()` as opposed to the syntactic sugar for various reasons so even when they will introduce this feature into the language if ever I doubt it makes sense to change existing types that has a single method just to take advantage of this feature but in some domains or DSL it can make things more readable.

**Reply by HaloFour** [reply]
I personally think that it's ok for the compiler to be opinionated here.

**Comment by HaloFour** [comment]
> With my code style often taking a codebase to the point of mostly single-method interfaces

Orthogonal but related: #2517

I'd love to see them considered together.  I think it would make it a lot easier to work with single-method interfaces in general.

```cs
IGetCustomerQuery query = customerId => new Customer(customerId);
Customer customer = query("123");
```

**Comment by yaakov-h** [comment]
Would this lead to ambiguous syntax problems, or does the precedent of `Delegate` and friends already solve this?
